File: scala/sum-of-multiples/src/main/scala/SumOfMultiples.scala
object SumOfMultiples:
  def sum(factors: Set[Int], limit: Int): Int =
    var uniqueMul: Set[Int] = Set()

    for
      factor <- factors if factor != 0
      num <- (1 until limit) if num % factor == 0
    yield uniqueMul = uniqueMul + num

    uniqueMul.foldLeft(0)(_ + _)

File: elixir/sum-of-multiples/lib/sum_of_multiples.ex
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    factors
    |> Enum.reduce(MapSet.new(), fn factor, set ->
      cond do
        factor >= limit or factor == 0 ->
          set

        true ->
          factor..(limit - 1)
          |> Stream.filter(&(rem(&1, factor) == 0))
          |> Enum.into(set)
      end
    end)
    |> Enum.sum()
  end
end
